### PR TITLE
Add `chainerx.fabs` and `chainerx.sign`

### PR DESCRIPTION
**Pull request by kshitij12345**
Adds `fabs, sign`

References: #6423

### CODE REVIEW

File: chainerx_cc/chainerx/cuda/numeric.cuh [hunk 1]
```diff
@@ -92,5 +92,21 @@ __device__ inline T Sqrt(T x) {
...
+template <typename T>
+__device__ inline T Fabs(T x) {
+    return std::fabs(x);
+}
+
+__device__ inline cuda::Float16 Fabs(cuda::Float16 x) { return cuda::Float16{std::fabs(static_cast<float>(x))}; }
+
+template <typename T>
+__device__ inline T Sign(T x) {
+    return IsNan(x) ? x : ((T(0) < x) - (x < T(0)));
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply chainerx_cc/chainerx/cuda/numeric.cuh]: Ah I see, I will directly commit the suggestion. Thank You very much.

File: chainerx_cc/chainerx/numeric.h [hunk 2]
```diff
@@ -125,4 +125,22 @@ inline chainerx::Float16 Sqrt<chainerx::Float16>(chainerx::Float16 x) {
...
+    return std::fabs(x);
+}
+template <>
+inline chainerx::Float16 Fabs<chainerx::Float16>(chainerx::Float16 x) {
+    return Float16{std::fabs(static_cast<float>(x))};
+}
+
+template <typename T>
+inline T Sign(T x) {
+    return IsNan(x) ? x : ((T{0} < x) - (x < T{0}));
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply chainerx_cc/chainerx/numeric.h]: Ah I see, I will directly commit the suggestion. Thank You very much.

**Comment by kshitij12345** [comment]
Ouch, did not notice the conflicts, will fix them.

**Comment by kshitij12345** [comment]
@niboshi I had some trouble with `_fallback_workaround.py` shadowing the implementation. It would be nicer, if it's mentioned in the #6423 over somewhere, that you may or may not have to check there.

Also please review.

**Comment by hvy** [comment]
@kshitij12345 I updated the description in #6423 in the section of _How do you start implementing routines?_ to mention the fallback.

**Comment by kshitij12345** [comment]
@hvy Ah thank you.

File: chainerx_cc/chainerx/native/native_device/misc.cc [hunk 3]
```diff
@@ -15,6 +15,42 @@ namespace chainerx {
...
+            using T = typename decltype(pt)::type;
+            struct Impl {
+                void operator()(int64_t /*i*/, T x, T& out) { out = chainerx::Sign(x); }
+            };
+            Elementwise<const T, T>(Impl{}, x_cast, out);
+        });
+    }
+};
+
+CHAINERX_REGISTER_OP_NATIVE(SignOp, NativeSignOp);
```
Comment (niboshi) [inline chainerx_cc/chainerx/native/native_device/misc.cc]: We recently renamed the macro to `CHAINERX_NATIVE_REGISTER_OP`. Please rebase to master and fix macro names.

File: chainerx_cc/chainerx/cuda/cuda_device/misc.cu [hunk 4]
```diff
@@ -62,6 +62,50 @@ public:
...
+        CudaSetDeviceScope scope{device.index()};
+        const Array& x_cast = x.dtype() == out.dtype() ? x : x.AsType(out.dtype());
+        VisitNumericDtype(out.dtype(), [&](auto pt) {
+            using T = typename decltype(pt)::type;
+            Elementwise<const T, T>(SignImpl<T>{}, x_cast, out);
+        });
+    }
+};
+
+CHAINERX_REGISTER_OP_CUDA(SignOp, CudaSignOp);
```
Comment (niboshi) [inline chainerx_cc/chainerx/cuda/cuda_device/misc.cu]: Ditto. We renamed it to `CHAINERX_CUDA_REGISTER_OP`.

File: chainerx_cc/chainerx/native/native_device/misc.cc (hunk 3 above)
Comment (kshitij12345) [reply chainerx_cc/chainerx/native/native_device/misc.cc]: I guess I'll wait till the #6944 goes in.

File: chainerx_cc/chainerx/cuda/cuda_device/misc.cu (hunk 4 above)
Comment (kshitij12345) [reply chainerx_cc/chainerx/cuda/cuda_device/misc.cu]: I guess I'll wait till the #6944 goes in.

**Comment by hvy** [comment]
#6944 is now merged.

**Comment by kshitij12345** [comment]
Yeah sure. Thanks for the heads up.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Sorry, could you resolve conflict once again?

**Comment by niboshi** [comment]
Jenkins, test this please

File: chainerx_cc/chainerx/cuda/numeric.cuh [hunk 5]
```diff
@@ -43,6 +43,16 @@ __device__ inline float Arccosh(float x) { return std::acoshf(x); }
 
 __device__ inline cuda::Float16 Arccosh(cuda::Float16 x) { return cuda::Float16{std::acoshf(static_cast<float>(x))}; }
 
+template <typename T>
+__device__ inline T Sign(T x) {
+    return IsNan(x) ? x : static_cast<T>((T{0} < x) - (x < T{0}));
```
Comment (niboshi) [inline chainerx_cc/chainerx/cuda/numeric.cuh]: Perhaps this line should also be fixed similarly to `cuda::Float16` specialization
(Avoid implicit cast from `bool` to `int`)?

File: chainerx_cc/chainerx/numeric.h [hunk 6]
```diff
@@ -26,6 +26,16 @@ inline bool IsInf(chainerx::Float16 value) { return value.IsInf(); }
 inline bool IsInf(double value) { return std::isinf(value); }
 inline bool IsInf(float value) { return std::isinf(value); }
 
+template <typename T>
+inline T Sign(T x) {
+    return IsNan(x) ? x : static_cast<T>((T{0} < x) - (x < T{0}));
```
Comment (niboshi) [inline chainerx_cc/chainerx/numeric.h]: Ditto.

File: chainerx_cc/chainerx/cuda/numeric.cuh (hunk 5 above)
Comment (kshitij12345) [reply chainerx_cc/chainerx/cuda/numeric.cuh]: Makes sense. On it.

**Comment by niboshi** [comment]
Thanks. Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
LGTM
Thank you so much!

**Pull request merged**